From 49cbe7d14e0362c2ded2e31c83b75b795fc2e14b Mon Sep 17 00:00:00 2001
From: allen yan <yanwei@marvell.com>
Date: Fri, 31 Mar 2017 11:09:17 +0800
Subject: [PATCH 1568/2241] fix: uart: a3700: free the uart tx irq when
 shutdown the uart port

- The TX and RX interrupts are not all free in mvebu_uart_shutdown.
  So irq number for TX interrupts will be duplicated once the port
  starts up again.
- This patch freed both TX and RX interrupts if uart interrupt summary
  is not used.
- JIRA BUG: A3700-889.

Change-Id: Iae9b0f42160011362f691634c89e708c4e9ff83d
Signed-off-by: allen yan <yanwei@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/38298
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Hua Jing <jinghua@marvell.com>
---
 drivers/tty/serial/mvebu-uart.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/mvebu-uart.c b/drivers/tty/serial/mvebu-uart.c
index ebac09f..554dc2a 100644
--- a/drivers/tty/serial/mvebu-uart.c
+++ b/drivers/tty/serial/mvebu-uart.c
@@ -566,7 +566,12 @@ static void mvebu_uart_shutdown(struct uart_port *port)
 	writel(0, port->membase + REG_CTRL(uart_data));
 	writel(0, port->membase + uart_data->intr.ctrl_reg);
 
-	free_irq(port->irq, port);
+	if (uart_data->intr.irq_sum > 0)
+		free_irq(port->irq, port);
+	else {
+		free_irq(uart_data->intr.irq_tx, port);
+		free_irq(uart_data->intr.irq_rx, port);
+	}
 }
 
 static void mvebu_uart_baud_rate_set(struct uart_port *port, unsigned int baud)
-- 
2.7.4

